### PR TITLE
Resolve relative artifact uri when creating experiments for File and SqlAlchemy Stores

### PR DESCRIPTION
**Pull request by bali0019**
Signed-off-by: Jas Bali

## Related Issues/PRs

 #5110

## What changes are proposed in this pull request?

For File and SqlAlchemy stores, the relative artifact URI is resolved relative to the current working directory but is serialized as a relative artifact location to an experiment's `meta.yaml`. But this means, when a run is created under that experiment, it always logs artifacts for that run under the current working directory. This is not consistent logging since it will change based on the current working directory. This PR addresses the core issue of resolving the artifact path relative to the working directory and serializes the absolute path (for artifact location) to the experiment's `meta.yaml`, which ensures artifacts are always logged under that location.
## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

When an experiment is created using local tracking URI or SQL alchemy-compliant backend store and artifact location is passed as a relative path, it will always be resolved relative to the current working directory at the time of experiment creation

Components

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/store/tracking/test_file_store.py [hunk 1]
```diff
@@ -533,12 +534,15 @@ def test_create_experiment(self):
 
     def test_create_experiment_appends_to_artifact_uri_path_correctly(self):
         cases = [
-            ("path/to/local/folder", "path/to/local/folder/{e}"),
+            ("path/to/local/folder", "{cwd}/path/to/local/folder/{e}"),
             ("/path/to/local/folder", "/path/to/local/folder/{e}"),
-            ("#path/to/local/folder?", "#path/to/local/folder?/{e}"),
-            ("file:path/to/local/folder", "file:path/to/local/folder/{e}"),
+            ("#path/to/local/folder?", "{cwd}/#path/to/local/folder?/{e}"),
```
Comment (bali0019) [inline tests/store/tracking/test_file_store.py:539]: @dbczumar, I know we discussed this offline, but I just want to make sure about this change......if a relative path looks like a fragment/query param (which doesn't make sense for a "relative" URI), with this PR change, it will be resolved as a literal path. IDK if this test makes sense here, but we shouldn't allow such a relative path with fragment/query param to be passed in like this for artifact locations.

File: tests/store/tracking/test_file_store.py [hunk 2]
```diff
@@ -533,12 +534,15 @@ def test_create_experiment(self):
...
+            ("file:path/to/local/folder", "{cwd}/file:path/to/local/folder/{e}"),
             ("file:///path/to/local/folder", "file:///path/to/local/folder/{e}"),
-            ("file:path/to/local/folder?param=value", "file:path/to/local/folder/{e}?param=value"),
+            (
+                "file:path/to/local/folder?param=value",
+                "{cwd}/file:path/to/local/folder?param=value/{e}",
```
Comment (bali0019) [inline tests/store/tracking/test_file_store.py:544]: @dbczumar ..and here. Notice, how the query param is now resolved as a literal path too compared to the existing test. To be specific
this relative path is now resolved as `file:path/to/local/folder?param=value` is resolved as `{cwd}/file:path/to/local/folder?param=value/{e}`
versus existing assert of this `file:path/to/local/folder?param=value` to `file:path/to/local/folder/{e}?param=value`

Comment (bali0019) [reply tests/store/tracking/test_file_store.py:544]: if I don't parse that a literal and attempt a URI parsing for it, I can technically grab fragment and query params and pass those along to the absolute artifact URI.

File: tests/store/tracking/test_file_store.py (hunk 1 above)
Comment (dbczumar) [reply tests/store/tracking/test_file_store.py:539]: If the user supplies just a fragment or query string for their artifact root directory or artifact URI, we should interpret it as a local filesystem path. Your changes to the test expectations make sense :)

File: tests/store/tracking/test_file_store.py (hunk 2 above)
Comment (dbczumar) [reply tests/store/tracking/test_file_store.py:544]: This should be `file:{cwd}/path/to/local/folder?param=value/{e}` because the supplied artifact location is a file URI with a relative path. We need to resolve the path to an absolute path and substitute it in the URIs path component, instead of prepending the current working directory to the URI.

File: mlflow/utils/uri.py [hunk 3]
```diff
@@ -281,3 +282,21 @@ def dbfs_hdfs_uri_to_fuse_path(dbfs_uri):
...
+    :param local_uri: Relative or absolute path or local file uri
+
+    :return: a fully-formed absolute uri path
+    """
+    from mlflow.utils.file_utils import local_file_uri_to_path
+
+    if local_uri is not None and is_local_uri(local_uri):
+        artifact_path = local_file_uri_to_path(local_uri)
+        if not pathlib.Path(artifact_path).is_absolute():
+            return str(pathlib.Path.cwd().joinpath(local_uri))
```
Comment (dbczumar) [inline mlflow/utils/uri.py:301]: This should be `artifact_path`, right? If `local_uri` is a `file:` URI, the current logic won't work.

```suggestion
            return str(pathlib.Path.cwd().joinpath(artifact_path))
```

File: mlflow/utils/uri.py [hunk 4]
```diff
@@ -281,3 +282,21 @@ def dbfs_hdfs_uri_to_fuse_path(dbfs_uri):
...
+    resolves it to absolute path relative to current working directory.
+
+    :param local_uri: Relative or absolute path or local file uri
+
+    :return: a fully-formed absolute uri path
+    """
+    from mlflow.utils.file_utils import local_file_uri_to_path
+
+    if local_uri is not None and is_local_uri(local_uri):
+        artifact_path = local_file_uri_to_path(local_uri)
```
Comment (dbczumar) [inline mlflow/utils/uri.py:299]: Nit: Can we call this `local_path` instead?

```suggestion
        local_path = local_file_uri_to_path(local_uri)
```

File: mlflow/utils/uri.py [hunk 5]
```diff
@@ -281,3 +282,21 @@ def dbfs_hdfs_uri_to_fuse_path(dbfs_uri):
         )
 
     return _DBFS_FUSE_PREFIX + dbfs_uri[len(_DBFS_HDFS_URI_PREFIX) :]
+
+
+def resolve_local_uri_if_relative(local_uri):
```
Comment (dbczumar) [inline mlflow/utils/uri.py:287]:
```suggestion
def resolve_uri_if_local(local_uri):
```

**Review by dbczumar — COMMENTED** [review]
@bali0019 Awesome stuff! Makes sense to move forward once comments are addressed.

File: tests/artifacts/test_artifacts.py [hunk 6]
```diff
@@ -136,3 +139,65 @@ def test_load_image_invalid_image(run_with_text_artifact):
...
+    artifact_name = "test.txt"
+    artifacts_root_tmp = mkdir(tmp_path.joinpath(str(uuid.uuid4())))
+    test_artifact_path = artifacts_root_tmp.joinpath(artifact_name)
+    with open(test_artifact_path, "w") as handle:
+        handle.write("test")
+    artifact_return_type = namedtuple(
+        "artifact_return_type", ["tmp_path", "artifact_path", "artifact_name"]
+    )
+    yield artifact_return_type(artifacts_root_tmp, test_artifact_path, artifact_name)
+    shutil.rmtree(artifacts_root_tmp, ignore_errors=True)
```
Comment (harupy) [inline tests/artifacts/test_artifacts.py:155]: Do we need this line? pytest automatically cleans up temporary directories created by the `tmp_path` fixture.

File: tests/store/tracking/test_file_store.py [hunk 7]
```diff
@@ -568,7 +572,9 @@ def test_create_experiment_appends_to_artifact_uri_path_correctly(self):
                 fs = FileStore(tmp.path(), artifact_root_uri)
                 exp_id = fs.create_experiment("exp")
                 exp = fs.get_experiment(exp_id)
-                assert exp.artifact_location == expected_artifact_uri_format.format(e=exp_id)
+                assert exp.artifact_location == expected_artifact_uri_format.format(
+                    e=exp_id, cwd=str(Path.cwd())
```
Comment (harupy) [inline tests/store/tracking/test_file_store.py:576]:
```suggestion
                    e=exp_id, cwd=Path.cwd()
```

File: tests/artifacts/test_artifacts.py [hunk 8]
```diff
@@ -136,3 +139,65 @@ def test_load_image_invalid_image(run_with_text_artifact):
...
         mlflow.artifacts.load_image(artifact.uri)
+
+
+@pytest.fixture()
+def text_artifact(tmp_path):
+    artifact_name = "test.txt"
+    artifacts_root_tmp = mkdir(tmp_path.joinpath(str(uuid.uuid4())))
+    test_artifact_path = artifacts_root_tmp.joinpath(artifact_name)
+    with open(test_artifact_path, "w") as handle:
+        handle.write("test")
```
Comment (harupy) [inline tests/artifacts/test_artifacts.py:150]:
```suggestion
    test_artifact_path.write_text("test")
```

File: tests/artifacts/test_artifacts.py (hunk 6 above)
Comment (harupy) [inline tests/artifacts/test_artifacts.py:155]:
```suggestion
    return artifact_return_type(artifacts_root_tmp, test_artifact_path, artifact_name)
```

Do we need this line? pytest automatically cleans up temporary directories created by the `tmp_path` fixture.

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 9]
```diff
@@ -609,7 +613,9 @@ def test_create_experiment_appends_to_artifact_uri_path_correctly(self):
                     )
                     exp_id = store.create_experiment(name="exp")
                     exp = store.get_experiment(exp_id)
-                    assert exp.artifact_location == expected_artifact_uri_format.format(e=exp_id)
+                    assert exp.artifact_location == expected_artifact_uri_format.format(
+                        e=exp_id, cwd=str(pathlib.Path.cwd())
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:617]:
```suggestion
                        e=exp_id, cwd=pathlib.Path.cwd()
```

File: mlflow/R/mlflow/tests/testthat/test-tracking-experiments.R [hunk 10]
```diff
@@ -249,9 +255,10 @@ test_that("experiment setting works", {
...
+  artifact_relative_path <- "artifact/location"
+  mlflow_set_experiment(experiment_name = "foo", artifact_location = artifact_relative_path)
   experiment <- mlflow_get_experiment()
-  expect_identical(experiment$artifact_location, "artifact/location")
+  expected_artifact_location <- sprintf("%s/%s", getwd(), artifact_relative_path)
+  expect_identical(experiment$artifact_location, expected_artifact_location)
   expect_identical(experiment$name, "foo")
-})
-
+})
```
Comment (harupy) [inline mlflow/R/mlflow/tests/testthat/test-tracking-experiments.R:264]: Can we insert a newline here?

File: mlflow/utils/uri.py [hunk 11]
```diff
@@ -281,3 +282,35 @@ def dbfs_hdfs_uri_to_fuse_path(dbfs_uri):
...
+
+
+def resolve_uri_if_local(local_uri):
+    """
+    if `local_uri` is passed in as a relative local path, this function
+    resolves it to absolute path relative to current working directory.
+
+    :param local_uri: Relative or absolute path or local file uri
+
+    :return: a fully-formed absolute uri path
```
Comment (dbczumar) [inline mlflow/utils/uri.py:294]: This return docstring doesn't quite seem accurate. It's either a URI with an absolute path component or an absolute filesystem path.

File: mlflow/utils/uri.py [hunk 12]
```diff
@@ -281,3 +282,35 @@ def dbfs_hdfs_uri_to_fuse_path(dbfs_uri):
...
+        cwd = pathlib.Path.cwd()
+        local_path = local_file_uri_to_path(local_uri)
+        if not pathlib.Path(local_path).is_absolute():
+            if scheme == "":
+                return str(cwd.joinpath(local_path))
+            local_uri_split = urllib.parse.urlsplit(local_uri)
+            resolved_absolute_uri = urllib.parse.urlunsplit(
+                (
+                    local_uri_split.scheme,
+                    None,
```
Comment (dbczumar) [inline mlflow/utils/uri.py:309]: Hmmm...`is_local_uri` appears to be broken for the case where `local_uri` is a `file://` URI containing a hostname that isn't localhost, e.g. `file://<some IP address>/path/to/file`. As a result, passing `None` for the host could cause problems here.

Can we update `is_local_uri` to check that the hostname component of the URI is empty before returning True?

File: mlflow/utils/uri.py [hunk 13]
```diff
@@ -281,3 +282,35 @@ def dbfs_hdfs_uri_to_fuse_path(dbfs_uri):
         )
 
     return _DBFS_FUSE_PREFIX + dbfs_uri[len(_DBFS_HDFS_URI_PREFIX) :]
+
+
+def resolve_uri_if_local(local_uri):
```
Comment (dbczumar) [inline mlflow/utils/uri.py:287]: Can we add some test cases for this method?

**Review by dbczumar — COMMENTED** [review]
@bali0019 This is looking great! Just a few small comments

File: mlflow/utils/uri.py (hunk 12 above)
Comment (dbczumar) [reply mlflow/utils/uri.py:309]: Can we add a test case for `is_local_uri` with a URI like `file://<some IP address>/path/to/file` to make sure it returns `False`?

File: mlflow/utils/uri.py [hunk 14]
```diff
@@ -17,8 +18,9 @@
 
 def is_local_uri(uri):
     """Returns true if this is a local file path (/foo or file:/foo)."""
-    scheme = urllib.parse.urlparse(uri).scheme
-    return uri != "databricks" and (scheme == "" or scheme == "file")
+    parsed_uri = urllib.parse.urlparse(uri)
+    scheme = parsed_uri.scheme
+    return uri != "databricks" and (scheme == "" or scheme == "file") and not parsed_uri.hostname
```
Comment (harupy) [inline mlflow/utils/uri.py:23]:
```suggestion
    if uri == "databricks":
        return False
    parsed_uri = urllib.parse.urlparse(uri)
    scheme = parsed_uri.scheme
    return (scheme == "" or scheme == "file") and not parsed_uri.hostname
```

File: mlflow/utils/uri.py [hunk 15]
```diff
@@ -176,11 +178,18 @@ def extract_and_normalize_path(uri):
...
+        return pathlib.Path(uri).as_posix()
+    return append_to_uri_path(uri, paths)
+
+
 def append_to_uri_path(uri, *paths):
     """
     Appends the specified POSIX `paths` to the path component of the specified `uri`.
 
-    :param uri: The input URI, represented as a string.
+    :param uri: The input URI" represented as a string.
```
Comment (harupy) [inline mlflow/utils/uri.py:192]: Can we revert this change?

File: tests/tracking/_tracking_service/test_utils.py [hunk 16]
```diff
@@ -145,7 +146,7 @@ def test_get_store_sqlalchemy_store(tmp_wkdir, db_type):
         store = _get_store()
         assert isinstance(store, SqlAlchemyStore)
         assert store.db_uri == uri
-        assert store.artifact_root_uri == "./mlruns"
+        assert store.artifact_root_uri == f"{Path.cwd().joinpath('mlruns')}"
```
Comment (harupy) [inline tests/tracking/_tracking_service/test_utils.py:149]:
```suggestion
        assert store.artifact_root_uri == str(Path.cwd().joinpath('mlruns')})
```

File: tests/utils/test_uri.py [hunk 17]
```diff
@@ -521,3 +527,57 @@ def test_dbfs_hdfs_uri_to_fuse_path(uri, result):
...
     with pytest.raises(MlflowException, match="did not start with expected DBFS URI prefix"):
         dbfs_hdfs_uri_to_fuse_path(path)
+
+
+def test_resolve_uri_if_local():
+    cwd = pathlib.Path.cwd()
+    input_and_expected_uris = {
+        "file://myhostname/my/path": "file://myhostname/my/path",
+        "file:///my/path": "file:///my/path",
+        "file:my/path": f"file://{cwd}/my/path",
```
Comment (harupy) [inline tests/utils/test_uri.py:537]: Is `"file:my/path"` a valid URI?

File: tests/utils/test_uri.py [hunk 18]
```diff
@@ -521,3 +527,57 @@ def test_dbfs_hdfs_uri_to_fuse_path(uri, result):
...
+        assert resolve_uri_if_local(input_uri) == expected_uri
+
+
+def test_is_local_windows_path():
+    from unittest import mock
+
+    with mock.patch(
+        "platform.system",
+        return_value="windows",
+    ):
```
Comment (harupy) [inline tests/utils/test_uri.py:553]:
```suggestion
```

I think we can remove `mock.patch` and skip this test on windows.

File: tests/utils/test_uri.py [hunk 19]
```diff
@@ -521,3 +527,57 @@ def test_dbfs_hdfs_uri_to_fuse_path(uri, result):
...
+        "my/path": f"{cwd}/my/path",
+        "/home/my/path": "/home/my/path",
+        "dbfs://databricks/a/b": "dbfs://databricks/a/b",
+        "s3://host/my/path": "s3://host/my/path",
+    }
+    for input_uri, expected_uri in input_and_expected_uris.items():
+        assert resolve_uri_if_local(input_uri) == expected_uri
+
+
+def test_is_local_windows_path():
```
Comment (harupy) [inline tests/utils/test_uri.py:547]:
```suggestion
@pytest.mark.skipif(os.name != "nt", reason="This test only passes on Windows")
def test_is_local_windows_path():
```

File: tests/utils/test_uri.py [hunk 20]
```diff
@@ -521,3 +527,57 @@ def test_dbfs_hdfs_uri_to_fuse_path(uri, result):
...
+        assert not _is_local_windows_path("\\a\\b\\c")
+
+
+def test_strip_leading_slash_if_windows_path():
+    from unittest import mock
+
+    with mock.patch(
+        "platform.system",
+        return_value="windows",
+    ):
```
Comment (harupy) [inline tests/utils/test_uri.py:570]:
```suggestion
```

Same here

File: mlflow/utils/uri.py [hunk 21]
```diff
@@ -162,12 +170,57 @@ def extract_db_type_from_uri(db_uri):
...
+    """
+    Detects if platform system is windows and if `uri_or_path` is a windows path
+    """
+    parsed_uri = urllib.parse.urlparse(uri_or_path)
+    scheme = parsed_uri.scheme
+    return (
+        platform.system().lower() == "windows"
+        and not parsed_uri.netloc
+        and uri_or_path[0] in string.ascii_letters
+        and uri_or_path[0].lower() == scheme.lower()
```
Comment (harupy) [inline mlflow/utils/uri.py:183]: What does this line check?

File: mlflow/utils/uri.py [hunk 22]
```diff
@@ -162,12 +170,57 @@ def extract_db_type_from_uri(db_uri):
...
+        and uri_or_path[1:3] in [":", ":/", "://", ":\\"]
+    )
+
+
+def _strip_leading_slash_if_windows_path(local_path):
+    """
+    Removes leading slash, if platform system is windows and `local_path` is a windows path
+
+    # For Windows path, urllib parses it with an extra leading slash
```
Comment (harupy) [inline mlflow/utils/uri.py:192]: Is this behavior specific to Windows?

```python
>>> urllib.parse.urlparse("file:////a/b/c")
ParseResult(scheme='file', netloc='', path='//a/b/c', params='', query='', fragment='')
>>> urllib.parse.urlparse("file:////C:/a/b/c")
ParseResult(scheme='file', netloc='', path='//C:/a/b/c', params='', query='', fragment='')
```

File: mlflow/utils/uri.py [hunk 23]
```diff
@@ -162,12 +170,57 @@ def extract_db_type_from_uri(db_uri):
...
+    return (
+        platform.system().lower() == "windows"
+        and not parsed_uri.netloc
+        and uri_or_path[0] in string.ascii_letters
+        and uri_or_path[0].lower() == scheme.lower()
+        and uri_or_path[1:3] in [":", ":/", "://", ":\\"]
+    )
+
+
+def _strip_leading_slash_if_windows_path(local_path):
```
Comment (harupy) [inline mlflow/utils/uri.py:188]: Do we pass a URI to this function?

File: mlflow/utils/uri.py [hunk 24]
```diff
@@ -281,3 +351,35 @@ def dbfs_hdfs_uri_to_fuse_path(dbfs_uri):
...
+    """
+    from mlflow.utils.file_utils import local_file_uri_to_path
+
+    if local_uri is not None and is_local_uri(local_uri):
+        scheme = get_uri_scheme(local_uri)
+        cwd = pathlib.Path.cwd()
+        local_path = local_file_uri_to_path(local_uri)
+        if not pathlib.Path(local_path).is_absolute():
+            if scheme == "":
+                return str(cwd.joinpath(local_path))
```
Comment (harupy) [inline mlflow/utils/uri.py:373]: Can we always return a URI?

File: mlflow/utils/uri.py (hunk 22 above)
Comment (bali0019) [reply mlflow/utils/uri.py:192]: yes, only specific to windows.

File: tests/utils/test_uri.py (hunk 17 above)
Comment (bali0019) [reply tests/utils/test_uri.py:537]: yes, urllib can parse it as:
```
import urllib.parse

urllib.parse.urlparse("file:my/path")
```
output:
```
ParseResult(scheme='file', netloc='', path='my/path', params='', query='', fragment='')
```

File: mlflow/utils/uri.py (hunk 24 above)
Comment (bali0019) [reply mlflow/utils/uri.py:373]: I had a brief chat with @dbczumar on this, but making this change would modify the current behaviour in master branch, check this existing test: https://github.com/mlflow/mlflow/blob/master/tests/store/tracking/test_file_store.py#L537

File: mlflow/utils/uri.py (hunk 22 above)
Comment (bali0019) [reply mlflow/utils/uri.py:193]: ...because POSIX fs absolute path will always have a leading slash anyways, but on windows, a drive letter is needed and not a leading slash.

on posix system:
```
import pathlib
print(pathlib.Path("/a/b/c").is_absolute())
```
ouput:
`True`

File: mlflow/utils/uri.py [hunk 25]
```diff
@@ -162,12 +170,62 @@ def extract_db_type_from_uri(db_uri):
...
+        platform.system().lower() == "windows"
+        and not parsed_uri.netloc
+        and len(local_parsed_path) > 3
+        and local_parsed_path[0] == "/"
+        and local_parsed_path[1] in string.ascii_letters
+        and local_parsed_path[2:4] in [":", ":/", "://"]
+    ):
+        final_path = local_parsed_path[1:]
+        if _is_local_windows_path(final_path):
+            return final_path
```
Comment (dbczumar) [inline mlflow/utils/uri.py:211]: I'm concerned that this logic might be fairly brittle. I know we're trying to convert relative URI paths to absolute URIs paths, but I think it may be better to just do this for UNIX paths and `file://` URIs. How does that sound?

File: mlflow/utils/uri.py [hunk 26]
```diff
@@ -281,3 +356,35 @@ def dbfs_hdfs_uri_to_fuse_path(dbfs_uri):
...
+            if scheme == "":
+                return cwd.joinpath(local_path).as_posix()
+            local_uri_split = urllib.parse.urlsplit(local_uri)
+            resolved_absolute_uri = urllib.parse.urlunsplit(
+                (
+                    local_uri_split.scheme,
+                    None,
+                    cwd.joinpath(local_path).as_posix(),
+                    local_uri_split.query,
+                    local_uri_split.fragment,
```
Comment (harupy) [inline mlflow/utils/uri.py:386]: Is this for a URI that looks like `file:///foo/bar?p=1#fff`

File: tests/store/tracking/test_file_store.py [hunk 27]
```diff
@@ -724,21 +759,63 @@ def test_get_deleted_runs(self):
...
+                    platform.system().lower() == "windows"
+                    and expected_artifact_uri_format.startswith("file:")
+                ):
+                    cwd = f"/{cwd}"
+                    drive = f"{drive}/"
+                assert run.info.artifact_uri == expected_artifact_uri_format.format(
+                    e=exp_id, r=run.info.run_id, cwd=cwd, drive=drive
+                )
+
+    @pytest.mark.skipif(os.name != "nt", reason="This test only passes on Windows")
```
Comment (BenWilson2) [inline tests/store/tracking/test_file_store.py:783]: Should we use the same conditional logic for path parsing for the test exclusion? 
```
platform.system().lower() == "windows"
```

File: mlflow/utils/uri.py [hunk 28]
```diff
@@ -281,3 +287,45 @@ def dbfs_hdfs_uri_to_fuse_path(dbfs_uri):
...
+    if `local_uri` is passed in as a relative local path, this function
+    resolves it to absolute path relative to current working directory.
+
+    :param local_uri: Relative or absolute path or local file uri
+
+    :return: a fully-formed absolute uri path or an absolute filesystem path
+    """
+    from mlflow.utils.file_utils import local_file_uri_to_path
+
+    if local_uri is not None and is_local_uri(local_uri):
```
Comment (BenWilson2) [inline mlflow/utils/uri.py:303]: very clean. Nice!

File: tests/conftest.py [hunk 29]
```diff
@@ -106,7 +106,7 @@ def new_exception(msg, *_, **__):
         yield
 
 
-@pytest.fixture(autouse=True, scope="module")
+@pytest.fixture(autouse=True)
```
Comment (BenWilson2) [inline tests/conftest.py:109]: Do we want to do this? How much runtime overhead for the tests is changing the scope back to the default (`function`) here?

Comment (BenWilson2) [reply tests/conftest.py:109]: Never mind. This fixture is faster than a time traveling delorean.

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 30]
```diff
@@ -1901,7 +1968,16 @@ def test_search_attrs(self):
         filter_string = "attribute.status = 'KILLED'"
         assert self._search([e1, e2], filter_string) == []
 
-        filter_string = f"attr.artifact_uri = '{ARTIFACT_URI}/{e1}/{r1}/artifacts'"
+        if platform.system().lower() == "windows":
```
Comment (BenWilson2) [inline tests/store/tracking/test_sqlalchemy_store.py:1971]: did you want to use your new test utility here?

File: tests/tracking/_tracking_service/test_utils.py [hunk 31]
```diff
@@ -145,7 +147,10 @@ def test_get_store_sqlalchemy_store(tmp_wkdir, db_type):
         store = _get_store()
         assert isinstance(store, SqlAlchemyStore)
         assert store.db_uri == uri
-        assert store.artifact_root_uri == "./mlruns"
+        if platform.system().lower() == "windows":
```
Comment (BenWilson2) [inline tests/tracking/_tracking_service/test_utils.py:150]: new utility here?

File: tests/tracking/test_rest_tracking.py [hunk 32]
```diff
@@ -91,7 +94,10 @@ def test_create_get_search_experiment(mlflow_client):
     )
     exp = mlflow_client.get_experiment(experiment_id)
     assert exp.name == "My Experiment"
-    assert exp.artifact_location == "my_location"
+    if platform.system().lower() == "windows":
```
Comment (BenWilson2) [inline tests/tracking/test_rest_tracking.py:97]: new utility here as well?

File: tests/artifacts/test_artifacts.py [hunk 33]
```diff
@@ -136,3 +140,87 @@ def test_load_image_invalid_image(run_with_text_artifact):
...
+                artifacts_root_path.joinpath(
+                    run.info.run_id, "artifacts", text_artifact.artifact_name
+                )
+            ),
+            text_artifact,
+            run.info.run_id,
+        )
+
+
+def test_artifact_logging_resolution_works_with_non_root_working_directory(text_artifact):
```
Comment (BenWilson2) [inline tests/artifacts/test_artifacts.py:204]: TY! :D

File: tests/utils/test_uri.py [hunk 34]
```diff
@@ -521,3 +526,41 @@ def test_dbfs_hdfs_uri_to_fuse_path(uri, result):
...
+def test_resolve_uri_if_local_on_windows():
+    input_and_expected_uris = {
+        "my/path": "file://{cwd}/my/path",
+        "file://myhostname/my/path": "file://myhostname/my/path",
+        "file:///my/path": "file:///{drive}my/path",
+        "file:my/path": "file://{cwd}/my/path",
+        "/home/my/path": "file:///{drive}home/my/path",
+        "dbfs://databricks/a/b": "dbfs://databricks/a/b",
+        "s3://host/my/path": "s3://host/my/path",
+    }
```
Comment (harupy) [inline tests/utils/test_uri.py:565]: Can we use parametrize this test?

```
@pytest.mark.parametrize(("uri", "expected"), ...)
def test_...
```

File: tests/tracking/test_tracking.py [hunk 35]
```diff
@@ -722,23 +740,28 @@ def test_get_artifact_uri_uses_currently_active_run_id():
...
+    _assert_get_artifact_uri_appends_to_uri_path_component_correctly(
+        "/dirname/rootpa#th?", "file:///{drive}/dirname/rootpa/{run_id}/artifacts/{path}#th?"
+    )
+
+
+@pytest.mark.skipif(is_local_os_windows(), reason="This test fails on Windows")
+def test_get_artifact_uri_appends_to_local_path_component_correctly():
+    _assert_get_artifact_uri_appends_to_uri_path_component_correctly(
+        "/dirname/rootpa#th?", "{drive}/dirname/rootpa#th?/{run_id}/artifacts/{path}"
+    )
```
Comment (harupy) [inline tests/tracking/test_tracking.py:764]: So on windows, `"/dirname/rootpa#th?"` is resolved to a URI, on non-windows OS, `"/dirname/rootpa#th?"` is resolved to a path?

File: tests/store/tracking/test_file_store.py [hunk 36]
```diff
@@ -724,21 +755,60 @@ def test_get_deleted_runs(self):
...
+            ("#path/to/local/folder?", "{cwd}/#path/to/local/folder?/{e}/{r}/artifacts"),
+        ]
+        self._assert_create_run_appends_to_artifact_uri_path_correctly(cases)
+
+    def test_create_run_appends_to_artifact_uri_path_correctly(self):
+        cases = [
+            ("file:path/to/local/folder", "file://{cwd}/path/to/local/folder/{e}/{r}/artifacts"),
+            (
+                "file:///path/to/local/folder",
+                "file:///{drive}path/to/local/folder/{e}/{r}/artifacts",
```
Comment (harupy) [inline tests/store/tracking/test_file_store.py:799]: I think this looks confusing because Unix systems don't have a drive.

**Review by harupy — APPROVED** [review]
Left a few more comments! LGTM once they are addressed

File: tests/tracking/test_tracking.py (hunk 35 above)
Comment (bali0019) [reply tests/tracking/test_tracking.py:764]: yes, that is correct! that is the intended change in this PR.

File: tests/store/tracking/test_file_store.py (hunk 36 above)
Comment (bali0019) [reply tests/store/tracking/test_file_store.py]: done and also, converted all artifacts tests from cases dict to parameterized pytests. It was just painful to find which tests failed one by one since it fails on first assert from that dict entry.

File: mlflow/utils/uri.py (hunk 26 above)
Comment (dbczumar) [reply mlflow/utils/uri.py:327]: @bali0019 Is this a valid file URI? If so, shall we include the query and fragment in the windows case as well?

Comment (dbczumar) [reply mlflow/utils/uri.py:327]: @bali0019 Can you confirm? (I think the answer is yes)

File: mlflow/utils/uri.py [hunk 37]
```diff
@@ -281,3 +287,45 @@ def dbfs_hdfs_uri_to_fuse_path(dbfs_uri):
...
+                    )
+                return cwd.joinpath(local_path).as_posix()
+            local_uri_split = urllib.parse.urlsplit(local_uri)
+            resolved_absolute_uri = urllib.parse.urlunsplit(
+                (
+                    local_uri_split.scheme,
+                    None,
+                    cwd.joinpath(local_path).as_posix(),
+                    local_uri_split.query,
+                    local_uri_split.fragment,
```
Comment (dbczumar) [inline mlflow/utils/uri.py:327]: Do we need the if / else conditional for Windows here, or could we consolidate the logic? Seems like we can

Comment (bali0019) [reply mlflow/utils/uri.py:327]: I believe we need the if/else somewhere because we need a file URI for windows and a local path for Unix systems in this logic handling

File: mlflow/utils/uri.py (hunk 26 above)
Comment (bali0019) [reply mlflow/utils/uri.py:327]: yup, that is correct. it just assembles back the URI that preserves original fragment and query.
To be specific, this is the test that confirms the behavior:
https://github.com/mlflow/mlflow/pull/7670/files#diff-9d5ba8ebadf9e5e465e40ff075a78fe5e61dd32525d3f146b440834ad06eb7d5R1887

File: mlflow/utils/uri.py (hunk 37 above)
Comment (bali0019) [reply mlflow/utils/uri.py:327]: hmmm, actually i take that back :)
I think we can consolidate this

Comment (bali0019) [reply mlflow/utils/uri.py:327]: Ok, I think the current version of this method is what I can do best without convoluting that logic further. if you have any better idea, I am happy to amend it.

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @bali0019 ! Awesome stuff!

**Comment by dbczumar** [comment]
@BenWilson2 Let me know if you have additional feedback here

**Review by BenWilson2 — APPROVED** [review]
LGTM! Let's ship it!
